torch/csrc/jit/codegen/cuda/lower_allocation.cpp

```diff
@@ -145,7 +146,146 @@ class AllocationInserter : public kir::MutableIrVisitor {
         continue;
       }
-      alloc_dims.push_back(id->extent());
+      auto extent = id->extent();
+      // Use halo-extended extent if found
+      auto halo_extent = gpu_lower->haloInfo().getRootAxisInfo(id);
+      if (halo_extent.width() != 0) {
+        extent = ir_builder.addExpr(
+            extent, ir_builder.create<kir::Int>(halo_extent.width()));
+      }
+      alloc_dims.push_back(extent);
+    }
+
+    return alloc_dims;
+  }
+
+  // Get allocation extents of root axes with halo
+  //
+  // Allocation can be done with leaf IDs with halo as well, but
+  // allocation size could be larger than necessary.
+  //
+  // For example, suppose the shift offset of an axis is 1. When it is
+  // split by N, the halo size of the inner output is N+1. When the
+  // allocation only has the inner split output, the allocation size
+  // would be N+1. Suppose that ID is further split by M, the output
+  // extents would be N/M and M+1. The allocation size based on the
+  // leaves would be N/M*(M+1) or N+N/M, which is larger than N+1.
+  //
+  // This function tries to propagate back halo informatin to root
+  // axes to avoid inflating allocations. It fails when merged domains
+  // are split and only one of the split outputs is used for
+  // allocations since in such a case we can't un-merge and properly
+  // determine the extents of the merge inputs. If that would result
+  // in an exception, but it may be more reasonable to simply fall
+  // back to the leaf-based allocation.
+  //
+  // See the FusionShiftDoubleSplit test for an example case.
+  std::vector<kir::Val*> getNonGlobalAllocExprWithHalo(
+      TensorView* tv,
+      const std::vector<IterDomain*>& alloc_domains) {
+    std::vector<Val*> start_vals;
+    std::transform(
+        alloc_domains.begin(),
+        alloc_domains.end(),
+        std::back_inserter(start_vals),
+        [](IterDomain* dom) { return dom->as<Val>(); });
+
+    // Get all exprs involved in generating the allocation IDs
+    auto exprs = ExprSort::getExprs(tv->fusion(), start_vals);
+
+    // Get the halo extent if found
+    auto getExtent = [this](IterDomain* id) {
+      auto extent = gpu_lower->haloInfo().getExtent(id);
+      if (extent == nullptr) {
+        extent = id->extent();
+      }
+      return gpu_lower->lowerValue(extent);
+    };
+
+    std::unordered_map<IterDomain*, kir::Val*> known_extents;
+
+    // IterDomains that are allocated fully. For example, if an ID is
+    // split and only one of them is used for allocation, that's not
+    // considered full. Only full domains can be unmerged, which is
+    // needed to propagate back the halo information to root domains.
+    std::unordered_set<IterDomain*> full_domains;
+
+    for (auto alloc_domain : alloc_domains) {
+      known_extents.insert({alloc_domain, getExtent(alloc_domain)});
+      full_domains.insert(alloc_domain);
+    }
+
+    for (auto it = exprs.rbegin(); it != exprs.rend(); ++it) {
```

Comment:
I think it would be good to use backward visitor for this like index compute. Doesn't need to be in this PR, but would be good to have a TODO.

Reply:
Added as a TODO in a separate issue (#877)